packages/zoe/state.js

```diff
@@ -1,190 +1,259 @@
-      for (const offerHandle of offerHandles) {
-        if (activeOffers.has(offerHandle)) {
-          active.push(offerHandle);
-        } else {
-          inactive.push(offerHandle);
-        }
-      }
-      return harden({
-        active,
-        inactive,
-      });
-    },
-  });
+const makeTable = (validateFn, makeCustomMethodsFn = () => {}) => {
+  // The WeakMap that stores the records
+  const handleToRecord = makePrivateName();
 
-  // The adminState should never leave Zoe and should be closely held
-  const adminState = harden({
-    addInstallation: installation => {
-      const installationHandle = harden({});
-      installationToInstallationHandle.init(installation, installationHandle);
-      installationHandleToInstallation.init(installationHandle, installation);
-      return installationHandle;
-    },
-    getInstallation: installationHandle =>
-      installationHandleToInstallation.get(installationHandle),
-    addInstance: async (
-      instanceHandle,
-      instance,
-      installationHandle,
-      terms,
-      assays,
-    ) => {
-      instanceHandleToInstance.init(instanceHandle, instance);
-      instanceHandleToInstallationHandle.init(
-        instanceHandle,
-        installationHandle,
-      );
-      instanceHandleToTerms.init(instanceHandle, terms);
-      instanceHandleToAssays.init(instanceHandle, assays);
-      for (let i = 0; i < assays.length; i += 1) {
-        // we want to wait until the first call returns before moving
-        // onto the next in case it is a duplicate assay
-        // eslint-disable-next-line no-await-in-loop
-        await adminState.recordAssay(assays[i]);
-      }
-    },
-    getInstance: instanceHandle => instanceHandleToInstance.get(instanceHandle),
-    getInstallationHandleForInstanceHandle: instanceHandle =>
-      instanceHandleToInstallationHandle.get(instanceHandle),
-    getPurses: assays => assays.map(assay => assayToPurse.get(assay)),
-    recordAssay: async assay => {
-      if (!assayToPurse.has(assay)) {
-        const unitOpsP = E(assay).getUnitOps();
-        const labelP = E(assay).getLabel();
-        const purseP = E(assay).makeEmptyPurse();
-        const extentOpsDescP = E(assay).getExtentOps();
-
-        const [unitOps, label, purse, extentOpsDesc] = await Promise.all([
-          unitOpsP,
-          labelP,
-          purseP,
-          extentOpsDescP,
-        ]);
-
-        assayToDescOps.init(assay, unitOps);
-        assayToLabel.init(assay, label);
-        assayToPurse.init(assay, purse);
-        const { name, extentOpArgs = [] } = extentOpsDesc;
-        assayToExtentOps.init(assay, extentOpsLib[name](...extentOpArgs));
-      }
-      return harden({
-        unitOps: assayToDescOps.get(assay),
-        label: assayToLabel.get(assay),
-        purse: assayToPurse.get(assay),
-        extentOps: assayToExtentOps.get(assay),
+  const table = harden({
+    validate: validateFn,
+    create: (handle, record) => {
```

Comment:
Consider putting the `handle` parameter last an have it default to a fresh empty pass-by-presence identity

```js
create: (record, handle = harden({})) => {
```

I expect that most uses would prefer to just have the table create the handle.

Reply:
Good idea

Reply:
That change is in